### PR TITLE
Added ability to pass custom callbacks while adhering to callback con…

### PR DESCRIPTION
**Pull request by JSarsfield**
…figuration

## :microscope: Background

- Why is this change needed? Is there a related issue or a new feature to be added?
- Current implementation doesn't allow custom callbacks to be passed in trainer_config.
- Early_stopping requires adjusting patience and divergence_threshold to work alongside regularisation
## :crystal_ball: Key changes

- Explain the main changes introduced by this pull request for the reviewer.
- Added ability to pass custom callbacks to trainer_config which adhere to current checks and flags e.g. passing EarlyStopping callback requires early_stopping flag to equal True.

## :clipboard: Review Checklist
- [ Y] I have performed a self-review of my own code.
- [ -] I have commented my code, added docstrings and data types to function definitions.
- [ -] I have added pytests to check whether my feature / fix works.

Please make sure to follow our best practices in the [Contributing guidelines](https://github.com/ourownstory/neural_prophet/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: neuralprophet/utils.py [hunk 1]
```diff
@@ -806,35 +806,71 @@ def configure_trainer(
...
+    has_custom_callbacks = True if "callbacks" in config else False
 
     # Configure checkpointing
-    if checkpointing_enabled:
-        # Callback to access both the last and best model
-        checkpoint_callback = pl.callbacks.ModelCheckpoint(
-            monitor=early_stopping_target, mode="min", save_top_k=1, save_last=True
+    has_modelcheckpoint_callback = (
+        True
+        if has_custom_callbacks and any(isinstance(x, pl.callbacks.ModelCheckpoint) for x in config["callbacks"])
```
Comment (karl-richter) [inline neuralprophet/utils.py:814]: Use self-explanatory variable name
```suggestion
        if has_custom_callbacks and any(isinstance(callback, pl.callbacks.ModelCheckpoint) for callback in config["callbacks"])
```

File: neuralprophet/utils.py [hunk 2]
```diff
@@ -806,35 +806,71 @@ def configure_trainer(
...
-        callbacks.append(checkpoint_callback)
+    if checkpointing_enabled:
+        if not has_modelcheckpoint_callback:
+            # Callback to access both the last and best model
+            checkpoint_callback = pl.callbacks.ModelCheckpoint(
+                monitor=early_stopping_target, mode="min", save_top_k=1, save_last=True
+            )
+            callbacks.append(checkpoint_callback)
+        else:
+            checkpoint_callback = next(cb for cb in config["callbacks"] if isinstance(cb, pl.callbacks.ModelCheckpoint))
```
Comment (karl-richter) [inline neuralprophet/utils.py:829]:
```suggestion
            checkpoint_callback = next(callback for callback in config["callbacks"] if isinstance(callback, pl.callbacks.ModelCheckpoint))
```

File: neuralprophet/utils.py [hunk 3]
```diff
@@ -806,35 +806,71 @@ def configure_trainer(
...
     else:
         config["enable_checkpointing"] = False
         checkpoint_callback = None
 
     # Configure the progress bar, refresh every epoch
+    has_progressbar_callback = (
+        True
+        if has_custom_callbacks and any(isinstance(x, pl.callbacks.ProgressBar) for x in config["callbacks"])
```
Comment (karl-richter) [inline neuralprophet/utils.py:837]:
```suggestion
        if has_custom_callbacks and any(isinstance(callback, pl.callbacks.ProgressBar) for callback in config["callbacks"])
```

File: neuralprophet/utils.py [hunk 4]
```diff
@@ -806,35 +806,71 @@ def configure_trainer(
...
+        if not has_progressbar_callback:
+            prog_bar_callback = ProgressBar(refresh_rate=num_batches_per_epoch, epochs=config_train.epochs)
+            callbacks.append(prog_bar_callback)
     else:
         config["enable_progress_bar"] = False
 
     # Early stopping monitor
+    has_earlystopping_callback = (
+        True
+        if has_custom_callbacks and any(isinstance(x, pl.callbacks.EarlyStopping) for x in config["callbacks"])
```
Comment (karl-richter) [inline neuralprophet/utils.py:854]:
```suggestion
        if has_custom_callbacks and any(isinstance(callback, pl.callbacks.EarlyStopping) for callback in config["callbacks"])
```

**Review by karl-richter — APPROVED** [review]
Thanks for your contribution, I find your changes really valuable!
Sorry for the long review time, we are currently preparing some major refactoring.

**Review by ourownstory — APPROVED** [review]
Thank you for your contribution, it is welcome!

**Comment by ourownstory** [comment]
Great work @JSarsfield

**Pull request merged**